### PR TITLE
Support ES2019 Array.prototype.flatMap

### LINKED ISSUE #949: Support ES2019 Array.prototype.flatMap
Proposal: https://github.com/tc39/proposal-flatMap
MDN:https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap

### PR DESCRIPTION
**Pull request by midgleyc**
Closes #949.

I followed what I did in #1313 but for flatMap this time.

### CODE REVIEW

**Comment by rbri** [comment]
please run spotlessApply to have the code in the correct layout

**Comment by midgleyc** [comment]
Array-like objects already work to some degree. There are four total test files:
```
array-like-objects.js
array-like-objects-nested.js
array-like-objects-typedarrays.js
array-like-objects-poisoned-length.js
```

Of those, the base and poisoned length fail, and the others pass.

**Comment by rbri** [comment]
@midgleyc build failed

org.mozilla.javascript.tests.MozillaSuiteTest > runMozillaTest[6550, js=testsrc/tests/js1_5/Regress/regress-89443.js, opt=9] STANDARD_ERROR
    java.lang.StackOverflowError
    	at org.mozilla.javascript.optimizer.Optimizer.rewriteForNumberVariables(Optimizer.java)
    	at org.mozilla.javascript.optimizer.Optimizer.rewriteAsObjectChildren(Optimizer.java:409)
    	at org.mozilla.javascript.optimizer.Optimizer.rewriteForNumberVariables(Optimizer.java:393)
    	at org.mozilla.javascript.optimizer.Optimizer.rewriteAsObjectChildren(Optimizer.java:409)
    	at org.mozilla.javascript.optimizer.Optimizer.rewriteForNumberVariables(Optimizer.java:399)
    	at org.mozilla.javascript.optimizer.Optimizer.rewriteAsObjectChildren(Optimizer.java:409)
    	at org.mozilla.javascript.optimizer.Optimizer.rewriteForNumberVariables(Optimizer.java:399)
    	at org.mozilla.javascript.optimizer.Optimizer.rewriteAsObjectChildren(Optimizer.java:409)
    	at org.mozilla.javascript.optimizer.Optimizer.rewriteForNumberVariables(Optimizer.java:399)
    	at org.mozilla.javascript.optimizer.Optimizer.rewriteAsObjectChildren(Optimizer.java:409)
    	at org.mozilla.javascript.optimizer.Optimizer.rewriteForNumberVariables(Optimizer.java:399)
    	at org.mozilla.javascript.optimizer.Optimizer.rewriteAsObjectChildren(Optimizer.java:409)
    	at org.mozilla.javascript.optimizer.Optimizer.rewriteForNumberVariables(Optimizer.java:399)
    	at org.mozilla.javascript.optimizer.Optimizer.rewriteAsObjectChildren(Optimizer.java:409)
    	at org.mozilla.javascript.optimizer.Optimizer.rewriteForNumberVariables(Optimizer.java:399)
    	at org.mozilla.javascript.optimizer.Optimizer.rewriteAsObjectChildren(Optimizer.java:409)
    	at org.mozilla.javascript.optimizer.Optimizer.rewriteForNumberVariables(Optimizer.java:399)
    	at org.mozilla.javascript.optimizer.Optimizer.rewriteAsObjectChildren(Optimizer.java:409)

**Comment by midgleyc** [comment]
Two tests fail: regress-89443 and regress-98901, both with Stack Overflows. Unfortunately, when I run them alone (in Java 17) they both pass. Any idea why they might be failing on GitHub?

**Comment by midgleyc** [comment]
I rebased and the issue went away.

File: src/org/mozilla/javascript/NativeArray.java [hunk 1]
```diff
@@ -441,6 +445,9 @@ public Object execIdCall(
                 case Id_flat:
                     return js_flat(cx, scope, thisObj, args);
 
+                case Id_flat_map:
+                    return js_flat_map(cx, scope, thisObj, args);
```
Comment (tuchida) [inline src/org/mozilla/javascript/NativeArray.java:449]:
```suggestion
                case Id_flatMap:
                    return js_flatMap(cx, scope, thisObj, args);
```
I want the case to be the same as other method names.
https://github.com/mozilla/rhino/blob/90d3b44e91fd70e7291dc9572eacc0376afa7872/src/org/mozilla/javascript/NativeArray.java#L430-L431

**Review by tuchida — APPROVED** [review]
Others look good. Thanks!

**Comment by gbrail** [comment]
Looks good. Thanks for working on this!

**Pull request merged**